contracts/AdjudicationFramework.sol

```diff
@@ -308,133 +314,118 @@ contract AdjudicationFramework is BalanceHolder {
+        address arbitrator = propositions[question_id].arbitrator;
+        bytes32 realityEthResult = realityETH.resultFor(question_id);
+        require(realityEthResult == bytes32(uint256(1)), "Result was not 1");
 
-        if (num == token_reservations[resid].num) {
-            delete (token_reservations[resid]);
-        } else {
-            token_reservations[resid].num = token_reservations[resid].num - num;
+        if (propositions[question_id].isFrozen) {
+            countArbitratorFreezePropositions[arbitrator] = countArbitratorFreezePropositions[arbitrator] - 1;
         }
+        delete(propositions[question_id]);
 
-        return reserveTokens(num, price, nonce);
+        arbitrators[arbitrator] = false;
     }
 
-    function cancelTimedOutReservation(bytes32 resid) external {
-        uint256 age = block.timestamp - token_reservations[resid].reserved_ts;
-        require(age > TOKEN_RESERVATION_CLAIM_TIMEOUT, "Not timed out yet");
-        reserved_tokens = reserved_tokens - token_reservations[resid].num;
-        delete (token_reservations[resid]);
-    }
+    // When an arbitrator is listed for removal, they can be frozen given a sufficient bond
+    function freezeArbitrator(
+        bytes32 question_id
+    ) public {
+        require(propositions[question_id].proposition_type == PropositionType.REMOVE_ARBITRATOR, "Wrong Proposition type");
+        address arbitrator = propositions[question_id].arbitrator;
 
-    function executeTokenSale(
-        bytes32 resid,
-        uint256 gov_tokens_paid
-    ) external l1_forkmanager_only {
-        uint256 age = block.timestamp - token_reservations[resid].reserved_ts;
         require(
-            age > TOKEN_RESERVATION_BIDDING_PERIOD,
-            "Bidding period has not yet passed"
+            arbitrators[arbitrator],
+            "Arbitrator not allowlisted" // Not allowlisted in the first place
         );
+        require(!propositions[question_id].isFrozen, "Arbitrator already frozen");
 
-        uint256 num = token_reservations[resid].num;
-        uint256 price = token_reservations[resid].price;
-        uint256 cost = price * num;
-        require(gov_tokens_paid >= cost, "Insufficient gov tokens sent");
-        reserved_tokens = reserved_tokens - num;
-        payable(token_reservations[resid].reserver).transfer(num);
+        // Require a bond of at least the specified level
+        // This is only relevant if REALITY_ETH_BOND_ARBITRATOR_FREEZE is higher than REALITY_ETH_BOND_ARBITRATOR_REMOVE
+        require(realityETH.getBond(question_id) >= REALITY_ETH_BOND_ARBITRATOR_FREEZE, "Bond too low to freeze");
 
-        delete (token_reservations[resid]);
+        // TODO: Do we want this? Freeze can be prevented by changing the answer after the bond is submitted
+        // Otherwise need a history to prove that the answer was given earlier with this bond
+        require(realityETH.getBestAnswer(question_id) == bytes32(uint256(1)), "Best answer is not yes");
```

Comment:
I mean, we could allow you to pass in the history and check there's a sufficient bond in the history. (Or you can wrap the transaction to do both in the same call). It's just a bit weird if it goes 

Should we remove this arbitrator?

Yes: $1
No: $1000000
...then we freeze them on the basis of the "no" bond...

Reply:
good point!

But realityeth is build up so that one does not need to look up the whole history, right? Only looking for the last Yes amount would be sufficient, right? Is it hard to implement?

Reply:
Correct, it's not hard to implement.

Reply:
Fixed in 735bec5899a958417af1a473d3ddce988782f8f5